### ISSUE TITLE
PyOpenCL scan with struct values gives wrong results

### ISSUE THREAD

**Issue by mattwala** [jrprice/Oclgrind#134]
See [this gist](https://gist.github.com/mattwala/e6322456df45a566a103e594824bf4d2) for the problem. In essence, it's trying to do a scan over an array of structs, each of which has 3 elements. The output is different from what I would expect but Oclgrind doesn't report any error message.

This is a simplified example of what happens when you try to run Oclgrind on [boxtree](https://github.com/inducer/boxtree).

cc: @inducer

**Comment by jrprice** [comment]
I've reproduced the incorrect results, but when I run with Oclgrind and use the `--data-races` flag it reports several read-write race conditions, which could possibly be the cause of the errors.

I'll try and investigate further soon.

**Comment by jrprice** [comment]
If I disable compiler optimisations it seems to work though, so this probably is indeed a bug in either Oclgrind or LLVM.

Does the boxtree code also work if you disable optimisations (`--build-options -O0`)?

**Comment by mattwala** [comment]
Thanks for taking a look.

The scan in boxtree works when I disable optimizations (I'm getting other errors now, but I'm still sorting out what cause is.)

**Comment by jrprice** [comment]
OK, this is an LLVM bug. I've submitted [a patch that fixes it](https://reviews.llvm.org/D31569), but it might take a little while to land. I've checked that your test case produces the correct results with Oclgrind built against LLVM trunk with that patch applied.

I'll keep this bug open until that LLVM patch lands. I've also added [a test case that exposes this bug](https://github.com/jrprice/Oclgrind/commit/05f5eee6e45c95cf64007f9259654ca7f1e9446b) to Oclgrind's test suite, marked as XFAIL for now.
